examples/filter_data.py

```diff
@@ -125,15 +122,33 @@ def main(
 
     with nullcontext() if rank == 0 else redirect_stdout(None):
         if args.query_index and args.filter == "attribution":
-            print("loading index...")
-            query_index = (
-                load_index(args.query_index)
-                    .with_format("torch")
-                    .map(normalize_batch, batched=True, batch_size=512)      
-            )
+            print("loading query index...")
+            query_index = load_gradient_dataset(args.query_index).with_format("torch")
+            
+            # Calculate mean gradient in batches to avoid memory issues
+            print("Computing query mean gradient...")
+            mean_gradient = None
+            batch_size = 512
+            
+            for i in tqdm(range(0, len(query_index), batch_size), desc="Computing query mean"):
+                batch = query_index.select(range(i, min(i + batch_size, len(query_index))))
+                batch_gradients = torch.stack([g for g in batch["gradients"]])
+                
+                # Normalize gradients
+                batch_gradients = F.normalize(batch_gradients, dim=1)
+                
+                if mean_gradient is None:
+                    mean_gradient = batch_gradients.mean(0)
+                else:
+                    mean_gradient = (mean_gradient * i + batch_gradients.sum(0)) / (i + len(batch))
+                
+                # Clear memory
+                del batch_gradients
+            
+            query = mean_gradient.cuda()
+            del query_index, mean_gradient
+            torch.cuda.empty_cache()
```

Comment:
Many people in the alignment community add the `torch.cuda.empty_cache()` line when they're having memory leaks they don't know how to fix, but it doesn't fix the leaks or achieve anything in most contexts. They are not setting a good example, plz do not adopt 😅 

My understanding is that `torch.cuda.empty_cache()` releases VRAM reserved by PyTorch that it doesn't need anymore, so if you are simultaneously running other processes that need VRAM it can be helpful.